homeassistant/components/recorder/history.py

```diff
@@ -245,31 +245,27 @@ def _get_states_with_session(
     # last recorder run started.
     query = session.query(*QUERY_STATES)
 
-    most_recent_states_by_date = session.query(
-        States.entity_id.label("max_entity_id"),
-        func.max(States.last_updated).label("max_last_updated"),
-    ).filter(
-        (States.last_updated >= run.start) & (States.last_updated < utc_point_in_time)
```

Comment:
Did you test query time with leaving `(States.last_updated >= run.start) &` in the innermost query and the new one on the whole query?

Reply:
I've run a couple of variants through the query optimizer for this test, as it was my production system that cokes on the current queries. The current queries run at around 0.005 seconds (in comparison, the existing version runs for days on an AMD Ryzen 7 machine).

Reply:
Explain for the current query in this PR

![image](https://user-images.githubusercontent.com/195327/135111925-65b0efcd-dad6-4a64-b2aa-5b41fc509674.png)

Reply:
I've tested these version on SQLite, database size 15GB, around 65k rows in the `states` table, so around 3x speedup with only the innermost table

Version 1, execution time ~250ms
```sql
SELECT states.domain AS states_domain,
       states.entity_id AS states_entity_id,
       states.state AS states_state,
       states.attributes AS states_attributes,
       states.last_changed AS states_last_changed,
       states.last_updated AS states_last_updated
FROM states
INNER JOIN
  (SELECT max(states.state_id) AS max_state_id
   FROM states
   WHERE states.last_updated < '2021-09-25 06:04:59.999999'
   GROUP BY states.entity_id) AS anon_1 ON states.state_id = anon_1.max_state_id
   WHERE states.last_updated >= '2021-08-25 22:22:49'
```

Version 2, execution time ~750ms
```sql
SELECT states.domain AS states_domain,
       states.entity_id AS states_entity_id,
       states.state AS states_state,
       states.attributes AS states_attributes,
       states.last_changed AS states_last_changed,
       states.last_updated AS states_last_updated
FROM states
INNER JOIN
  (SELECT max(states.state_id) AS max_state_id
   FROM states
   WHERE states.last_updated >= '2021-08-25 22:22:49'
   AND states.last_updated < '2021-09-25 06:04:59.999999'
   GROUP BY states.entity_id) AS anon_1 ON states.state_id = anon_1.max_state_id
```